WIP: Fix definition of RawExtension in OpenAPI schema

luksa opened the pull request:
Currently, the `RawExtension` definiton in the OpenAPI schema is wrong, as it doesn't allow for arbitrary values to be passed in fields with the `RawExtension` type. 

This is the current definition:
```
   "io.k8s.apimachinery.pkg.runtime.RawExtension": {
    "description": "RawExtension is used to hold extensions in external versions.\n\nTo use this, make a field which has RawExtension as its type in your external, versioned struct, and Object in your internal struct. You also need to register your various plugin types.\n\n// Internal package: type MyAPIObject struct {\n\truntime.TypeMeta `json:\",inline\"`\n\tMyPlugin runtime.Object `json:\"myPlugin\"`\n} type PluginA struct {\n\tAOption string `json:\"aOption\"`\n}\n\n// External package: type MyAPIObject struct {\n\truntime.TypeMeta `json:\",inline\"`\n\tMyPlugin runtime.RawExtension `json:\"myPlugin\"`\n} type PluginA struct {\n\tAOption string `json:\"aOption\"`\n}\n\n// On the wire, the JSON will look something like this: {\n\t\"kind\":\"MyAPIObject\",\n\t\"apiVersion\":\"v1\",\n\t\"myPlugin\": {\n\t\t\"kind\":\"PluginA\",\n\t\t\"aOption\":\"foo\",\n\t},\n}\n\nSo what happens? Decode first uses json or yaml to unmarshal the serialized data into your external MyAPIObject. That causes the raw JSON to be stored, but not unpacked. The next step is to copy (using pkg/conversion) into the internal struct. The runtime package's DefaultScheme has conversion functions installed which will unpack the JSON stored in RawExtension, turning it into the correct object type, and storing it in the Object. (TODO: In the case where the object is of an unknown type, a runtime.Unknown object will be created and stored.)",
    "required": [
     "Raw"
    ],
    "properties": {
     "Raw": {
      "description": "Raw is the underlying serialization of this object.",
      "type": "string",
      "format": "byte"
     }
    }
   },
```

**What this PR does / why we need it**:
This PR changes the OpenAPI definition of RawExtension so that any object, array or primitive is accepted. This is the new definition:
```
   "io.k8s.apimachinery.pkg.runtime.RawExtension": {
    "description": "Any value (any object, array or primitive)"
   },
```

**Which issue(s) this PR fixes** 
Fixes https://github.com/kubernetes/kubernetes/issues/55890

**Special notes for your reviewer**:
This PR requires a PR in kube-openapi: https://github.com/kubernetes/kube-openapi/pull/25
I've already vendored the required changes into this PR, even though I'm sure that's not the proper way to do it. I guess I should get the other PR merged first, then properly vendor in the whole kube-openapi in a separate commit, right? Please advise.

**Release note**:
```release-note
NONE
```

deads2k commented:
/assign mbohlool 
/unassign

roycaihw commented:
/sig api-machinery

dims commented:
/ok-to-test

mengqiy commented:
/assign @apelisse

apelisse commented:
> I've already vendored the required changes into this PR, even though I'm sure that's not the proper way to do it. I guess I should get the other PR merged first, then properly vendor in the whole kube-openapi in a separate commit, right? Please advise.

Yeah, you will need the other one merged first, and re-vendor with the proper hash or what not. But it's kind of cool to see how it's used.

lavalamp commented:
cc @yliaog

smarterclayton commented:
RawExtension shouldn’t show up in the output of openapi (I can’t tell from this issue if that’s the intent).  It should mark the parent field “object”

smarterclayton commented:
(type object, that is

mbohlool commented on staging/src/k8s.io/apimachinery/pkg/runtime/types.go:
```diff
@@ -99,6 +104,16 @@ type RawExtension struct {
+		Schema: spec.Schema{
+			SchemaProps: spec.SchemaProps{
+				Description: "Any value (any object, array or primitive)",
+			},
```
I would add a `"type": "object"` here to be more explicit.

luksa replied:
Yes, I've already changed that (but have yet to update the PR). My initial understanding was that `RawExtension` was used for any type, not just for objects.

mbohlool commented:
While having `RawExtension` defined like that (and I think it needs a `"type": "object"` there to be explicit) I agree with @smarterclayton that having the parent objects fixed is a better solution. However if that prove to be hard to do, I am OK with this as it is valid OpenAPI v2 spec. @smarterclayton WDYT?

luksa commented:
I've now changed it according to @smarterclayton's comments, but still need to figure out how to get kube-openapi to handle references to RawExtension properly without having the RawExtension type references in kube-openapi (see https://github.com/luksa/kube-openapi/commit/65c6a671cc6b209d4af7f939589beb2468c16440#diff-5e7f2c228518b5fb84f27a2cbb8986b0R114 )

mbohlool commented:
@luksa Have you update the PR? I don't see new code.

luksa commented:
I have now.. sorry, this is currently a mess, as I was experimenting. I'm pushing just so you can take a look. I'll clean it up when we decide what the proper way to solve this is.

luksa commented on vendor/k8s.io/kube-openapi/pkg/common/common.go:
```diff
@@ -90,27 +90,28 @@ type Config struct {
+	"boolean":                                      {"boolean", ""},
+	"[]byte":                                       {"string", "byte"}, // base64 encoded characters
+	"interface{}":                                  {"object", ""},
+	"k8s.io/apimachinery/pkg/runtime.RawExtension": {"object", ""},
```
I don't like having `RawExtension` referenced in kube-openapi, since `RawExtension` isn't part of it, but I don't see a simple way of getting around this. Any ideas?

mbohlool replied:
what if we add `common.GetOpenAPITypeFormat` to `common.Config`? We can then customize it in main repo and pass it using the config struct.

luksa replied:
AFAICT, `common.Config` isn't used in the generator (`generators/openapi.go`).

mbohlool replied:
Hmm. Good point. Then they need to be a customization list (in addition to the list in `common.GetOpenAPITypeFormat`) either coming from command line or in [this file](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/code-generator/cmd/openapi-gen/main.go). I would prefer the later. wdyt?

pmorie replied:
I would support loading from the file @mbohlool referenced for now; we can load from a CLI flag in the future if we need to.

luksa commented:
@mbohlool So, registering the OpenAPITypeFormat for RawExtension in the generator itself (here: https://github.com/kubernetes/kubernetes/pull/56434/commits/3423282d0fe32fe2e118dd4856d1e4f7881af640#diff-f927293330f1ed9a49260a2147de4849R52) isn't problematic, but using the same mechanism elsewhere is problematic. 

I've tried to go the route you suggested previously ("what if we add common.GetOpenAPITypeFormat to common.Config? We can then customize it in main repo and pass it using the config struct."), but doesn't seem right to me. The problem is that things like the Service Catalog API server would have to register the OpenAPITypeFormat for RawExtension themselves instead of APIMachinery doing that automatically. Devs will forget to do that (or won't even know they need to), so the generated OpenAPI schema will include an incorrect definition of RawExtension. 

The only correct place to define the OpenAPITypeFormat for RawExtension is in k8s.io/apimachinery, since that's where RawExtension is defined. But, APIMachinery doesn't create the OpenAPI Config, so it can't do that at all. 

I'm starting to think including the RawExtension definition in the schema is an acceptable tradeoff over what @smarterclayton suggested.

mbohlool commented:
@luksa what if we add some kind of tag to RawExtension definition?

```go
// ...
// +k8s:openapi-gen:schemaTypeFormat=object,
```

luksa commented:
Yeah, that could work, but would require us to look up the type with reflection and obtain that tag. IIRC, the code currently never needs to look up the target type. For that reason, I dismissed this idea previously. But if you don't see a problem with that, I'll see tomorrow if I can make that work.  Thanks.

mbohlool commented:
Why does it need reflection? the generator step can also just generate `customSchemaTypeFormatMap` (from your commit) in addition to `GetOpenAPIDefinitions` and we can pass that map in `common.Config` to builder.

pmorie commented:
@luksa looks like verify job failed

luksa commented:
@mbohlool PTAL. I've introduced ˙+k8s:openapi-gen:schema-type-format˙ as you suggested. It makes everything much simpler. 

There is one big general problem with changing the RawExtension definition: it requires a change to `k8s.io/kube-openapi/pkg/util/proto/document.go`, which currently does not like OpenAPI object definitions that look like a map with arbitrary values. This is fixed by the following commit: https://github.com/kubernetes/kubernetes/pull/56434/commits/9fb3a8889167482366a3b71352e30e67de93aeeb. Current clients will therefore break with the following error:
```
invalid object doesn't have additional properties
```

Obviously, this change needs to go into kube-openapi and into clients before the OpenAPI schema in the server changes. AFAIK, every kubectl version must stay compatible with servers up to 2 subsequent versions. If that's the case, how do we introduce this change? Assuming the change in `document.go` goes into kubectl v1.10, we can merge the RawExtension definition change no earlier than v1.12.

luksa commented:
@pmorie yeah, godep verification fail, because this PR currently also includes a change to kube-openapi, which I'll move there when everyone agrees this PR is the way to go.

pmorie commented:
@luksa got it

mbohlool commented:
@luksa without looking at the code yet, is it possible to add additional property with type "object" for maps with arbitrary values?

mbohlool commented on vendor/k8s.io/kube-openapi/pkg/generators/openapi.go:
```diff
@@ -411,6 +412,19 @@ func (g openAPITypeWriter) generate(t *types.Type) error {
 
+func getOpenAPITypeFormat(t *types.Type) (string, string) {
+	schemaTypeFormatTag := types.ExtractCommentTags("+", t.CommentLines)[tagSchemaTypeFormat]
+	for _, val := range schemaTypeFormatTag {
```
there should be only one of these tags. If there is more than one, the generator should fail with an error message.

luksa replied:
Fixed.

mbohlool commented:
Also which object has the additionalProperty problem?

luksa commented:
@mbohlool 
> without looking at the code yet, is it possible to add additional property with type "object" for maps with arbitrary values?

Not sure what you mean, but the code in `document.go` treats every OpenAPI definition, which includes `type:object` as a map. And it then expects the definition to include an `additionalProperties` field.

The correct way to define an arbitrary object (an object that is allowed to contain any field) is to include only `type: object` in its definition (and not `properties` or `additionalProperties`). 

Actually, this reminds me. The definitions of the k8s API objects in the OpenAPI schema are not strictly correct at the moment. They should all include the `type: object` field (but they currently only contain `properties`, which, according to the OpenAPI specification, does _not_ imply that they are objects). Take `io.k8s.api.core.v1.PodSpec` for example:

```
   "io.k8s.api.core.v1.PodSpec": {
    "description": "PodSpec is a description of a pod.",
    "required": [
     "containers"
    ],
    "properties": {
     "activeDeadlineSeconds": {...},
     ...
    }
   },
```

According to the OpenAPI spec, the following Pod manifest is perfectly valid against the schema currently served by k8s:
```
apiVersion: v1
kind: Pod
spec: "I'm a string, not an object"
```

More info about this: https://stackoverflow.com/questions/47374980/schema-object-without-a-type-attribute-in-swagger-2-0

So, all the OpenAPI schema definitions that are generated from structs, should also include the `type:object` property. This will, however, break kubectl again (in the same way as my RawExtension change does).  I worked on this in December, but then forgot about it. I'll see if I can finish it this week. 


> Also which object has the additionalProperty problem?
One of them is `io.k8s.api.apps.v1.ControllerRevision.data`.

This is what happens, when you use an older kubectl against a server patched with this PR:
```
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"9", GitVersion:"v1.9.1", GitCommit:"3a1c9449a956b6026f075fa3134ff92f7d55f812", GitTreeState:"clean", BuildDate:"2018-01-04T11:52:23Z", GoVersion:"go1.9.2", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"10+", GitVersion:"v1.10.0-alpha.0.2186+a95472eb457c49-dirty", GitCommit:"a95472eb457c49bd04394266a988e05c0f801f5e", GitTreeState:"dirty", BuildDate:"2018-01-16T08:46:53Z", GoVersion:"go1.9.2", Compiler:"gc", Platform:"linux/amd64"}

$ kubectl explain po
error: SchemaError(io.k8s.api.apps.v1.ControllerRevision.data): invalid object doesn't have additional properties
```

pmorie commented:
@luksa, @mbohlool:

@liggitt and I were discussing what the best way forward on this today; we came up with this plan of attack:

- Extract the kube-openapi changes into a separate PR, which we can merge into master and cherry-pick back into the 1.9 release branch
- Keep the other changes in this PR, so we can ship them in 1.10, and give folks a chance to update their older kubectl binaries first so we don't break them

WDYT?

luksa committed:
TEMP: Cherry-pick PR https://github.com/kubernetes/kube-openapi/pull/85 (server-side kube-openapi schema fixes)

luksa committed:
Fix RawExtension definition in OpenAPI schema

luksa commented:
I've now split the changes into client-side and server-side. I've also fixed the missing `type:object` OpenAPI schema problem. PTAL

I haven't yet submitted PRs against kube-openapi (I think it's best if you do an initial review here), but you can see all the changes here:
- client-side: https://github.com/luksa/kube-openapi/commits/openapi_schema_fixes_clientside
- server-side: https://github.com/luksa/kube-openapi/commits/openapi_schema_fixes_serverside

I've tested my changes in the following scenarios:
- 1.9 client against 1.10+myPR API server (obviously does not work)
- 1.10+myPr client against 1.9 API server (works)
- 1.10+myPR client against 1.10+myPR API server (works)

@pmorie @liggitt I am a bit apprehensive about including the client-side changes into 1.9. I'd feel much more comfortable if they were enabled only when the client is talking to a 1.10 server, but I haven't looked at what that would entail. Note that PR https://github.com/kubernetes/kubernetes/pull/57104 and https://github.com/kubernetes/kube-openapi/pull/25 would also need to be merged into 1.9.

apelisse commented on staging/src/k8s.io/apimachinery/pkg/util/strategicpatch/types.go:
```diff
@@ -74,6 +74,10 @@ func (item *kindItem) VisitMap(schema *openapi.Map) {
 	item.err = errors.New("expected kind, but got map")
 }
 
+func (item *kindItem) VisitArbitrary(schema *openapi.Arbitrary) {
```
Are we sure about that? Shouldn't Arbitrary accept anything?

apelisse commented on staging/src/k8s.io/apimachinery/pkg/util/strategicpatch/types.go:
```diff
@@ -141,6 +145,10 @@ func (item *sliceItem) VisitMap(schema *openapi.Map) {
 	item.err = errors.New("expected slice, but got map")
 }
 
+func (item *sliceItem) VisitArbitrary(schema *openapi.Arbitrary) {
```
Same question as above.

apelisse reviewed (commented):
Two questions aside, looks good to me!

liggitt commented:
> @pmorie @liggitt I am a bit apprehensive about including the client-side changes into 1.9. I'd feel much more comfortable if they were enabled only when the client is talking to a 1.10 server, but I haven't looked at what that would entail.

the point is that we have to stage this over two releases, so we need the client side fixes in a separate PR. if the client side fixes are small, we could consider picking them back to 1.9.

luksa commented:
@liggitt Yeah, I know the client-side changes need to go in at least one version before the server-side ones. The thing is that the client-side changes are not trivial. I'd like to see a few people use a patched 1.9 version for at least a week to see if the patch breaks anything. If the changes looks reasonable, I'll go ahead and cherry pick them into a 1.9 branch.

luksa commented:
I've been thinking about the smallest possible change to 1.9 that would allow a 1.9 client to be compatible with the RawExtension OpenAPI schema change set to be included in 1.10. Here it is: https://github.com/luksa/kubernetes/commit/c189fce9f4db82149b45c17b607e1e356af69dee

Instead of reporting an error, the client will now treat OpenAPI definitions that are `type:object` but don't have `additionalProperties` (this is how the `RawExtension` type will look like in 1.10), as a `map[string]string`. The 1.9 API server doesn't return any such definition (if it did, the client would report an error), so there's no way that this change would break the client when used with a 1.9 or earlier API server.

The 1.10 API server _will_ return such object definitions. A 1.9 client connecting to a 1.10 server would treat RawExtension-type fields as `map[string]string` instead of `map[string]anyType`, which is not really a problem, since we can reasonably expect users to use the 1.10 client if they want to use the service catalog. Actually, they could still use the 1.9 client as long as they only pass in string values in things like `ServiceInstance.spec.properties`. 

I propose we introduce the other changes (related to the missing `type:object` in all object definitions that describe structs) in version 1.10 and 1.11 (client-side changes in 1.10, server-side changes in 1.11). 

WDYT? @mbohlool @pmorie @liggitt @apelisse

liggitt commented:
> Instead of reporting an error, the client will now treat OpenAPI definitions that are type:object but don't have additionalProperties (this is how the RawExtension type will look like in 1.10), as a `map[string]string`

It seems like the client should either ignore the field or treat it properly. Expecting arbitrary values to be strings seems equally broken.

> The 1.10 API server will return such object definitions. A 1.9 client connecting to a 1.10 server would treat RawExtension-type fields as map[string]string instead of map[string]anyType, which is not really a problem, since we can reasonably expect users to use the 1.10 client if they want to use the service catalog. Actually, they could still use the 1.9 client as long as they only pass in string values in things like ServiceInstance.spec.properties.

It's unclear to me what breaks in kubectl when the openapi spec is fixed. Is it just client-side validation?

cc @apelisse @pwittrock @mengqiy

luksa commented:
> It's unclear to me what breaks in kubectl when the openapi spec is fixed. Is it just client-side validation?

Sadly, what breaks is the client-side parsing of the OpenAPI schema, which means it affects all users, not just those that use the service catalog stuff. For example, creating a simple pod breaks:
```
$ kubectl apply -f pod.yaml 
error: SchemaError(io.k8s.api.apps.v1.ControllerRevision.data): invalid object doesn't have additional properties
```

> It seems like the client should either ignore the field or treat it properly. Expecting arbitrary values to be strings seems equally broken.

I agree it's equally broken, but fixing it properly is too big of a change for 1.9. I may be ok with ignoring the field, as you suggest, but I can't tell what the effect of that would be, apart from the fact that `kubectl explain` would no longer show such fields (e.g. `kubectl explain controllerrevision` wouldn't show the `data` field). Ignoring the field affects more than just validation.

I went ahead and implemented what you suggest (ignoring RawExtension fields instead of treating them as map[string]string): https://github.com/luksa/kubernetes/commit/e5878c21b230bd26c46e4d6be2643e69c2bdb622

liggitt commented:
> I may be ok with ignoring the field, as you suggest, but I can't tell what the effect of that would be, apart from the fact that kubectl explain would no longer show such fields (e.g. kubectl explain controllerrevision wouldn't show the data field). Ignoring the field affects more than just validation.

@apelisse @pwittrock @mengqiy can you weigh in on the effect of ignoring RawExtension fields in the client when parsing openapi specs? does this break apply?

luksa commented:
Nevermind. If we're not touching 1.9, we can get the proper client fix into 1.10 and the server-side changes into 1.11. I'll split this PR into two and we'll continue there.

nilebox commented:
> we can get the proper client fix into 1.10 and the server-side changes into 1.11

@luksa 1.10 has already been released, so it's gonna be 1.10 client -> 1.11 server I suppose.
Do you have plans on finishing this PR?

luksa commented:
@nilebox yes, I'll take time tomorrow to see what remains to be done.

mbohlool commented:
> Nevermind. If we're not touching 1.9, we can get the proper client fix into 1.10 and the server-side changes into 1.11. I'll split this PR into two and we'll continue there.

should we close this PR?

nilebox commented:
@luksa any progress on that? Do you have 2 split PRs as you planned? Please link them if you do.

luksa commented:
@nilebox the open-api client-side changes are already in master; the client now just needs https://github.com/kubernetes/kubernetes/pull/57104

Then, in the next version, https://github.com/kubernetes/kube-openapi/pull/85  needs to go into kube-openapi, and then this PR can be modified to pull in the server-side kube-openapi changes.
